### PR TITLE
[mono][interp] Disable inlining into bblocks that are detected as dead early during codegen

### PR DESCRIPTION
**Pull request by BrzVlad**
This PR addresses compilation overhead for methods containing patterns like:
```
if (false)
    Method1 ();
else if (false)
    Method2 ();
....
```
where MethodX are methods marked as aggressive inlining. This pattern is common in intrinsics related code. Before this PR we would inline all methods, together with all subcalls, even if the code path is obviously unreachable, leading to significant pressure on the compiler.

We do inlining, on the fly, in the IL code traversal stage. BBlock and constant folding optimizations are run afterwards once the full IL code has been processed. In order to be able to disable inlining in these dead bblocks, we implement some subset of cfolding to be applied as soon as we import IL code, so we can detect dead bblocks before we actually generate their code. An alternative would be to move inlining as a later phase, after the standard full bblock/cfold optimization passes are run, but that is a more invasive approach.

This works by adding jump counters for each bblock, which are set before we actually start generating code for the IL. As we generate code we optimize out some forward branches and decrement the jump counter for the target bblocks. When we then get to emit code in a bblock that is no longer a jump target for any other instruction, if the bblock is also not a fall through bblock from the previous bblock, it means it is dead and no inlining will happen into it. Also, when processing branches in such dead bblocks, we additionally decrement the jump counters for the target_bb. This isn't able to detect dead code involving loops, but it should cover most scenarios in the BCL.

This makes `System.Numerics.Tensors.Tests` go from taking 300s with mempeak of 6GB to a run time of 100s with mempeak of 600MB.

### CODE REVIEW

**Comment by BrzVlad** [comment]
cc @stephentoub @tannergooding

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -673,24 +673,59 @@ handle_branch (TransformData *td, int long_op, int offset)
 	if (offset > 0)
 		init_bb_stack_state (td, target_bb);
 
+	if (td->cbb->no_inlining && long_op != MINT_CALL_HANDLER)
+		target_bb->jump_targets--;
```
Comment (tannergooding) [inline src/mono/mono/mini/interp/transform.c:677]: Just wondering, why not just always eliminate dead blocks early if possible?

Presumably even something like this is worth simplifying, particularly for larger block bodies:
```csharp
if (someExpressionThatIsTriviallyConstantFalse)
{
    // logic that can never be hit
}
else
{
    // logic that will always be hit
}
```

I'd expect it reduces memory overhead, improves overall throughput due to less nodes to traverse, and may even make other optimizations easier to identify.

Comment (tannergooding) [reply src/mono/mono/mini/interp/transform.c:677]: 👍, I'm not sure exactly what RyuJIT does for this scenario, but @EgorBo would likely know.

The general consideration comes in for code like this: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs,2007

Where, even completely ignoring inlining, there exists a large block of code that is always dead for Mono today and will always be dead for most non-xarch platforms.

This type of pattern is pretty common outside the BCL in perf critical code, and that is often some of the more complex/infrastructural code for frameworks/libraries.

Comment (BrzVlad) [reply src/mono/mono/mini/interp/transform.c:677]: Actually, what is worse in the code sample you provided is the existance of the `do/while` loop. So in addition to traversing all that code, we will still inline.

**Review by lewing — APPROVED** [review]

**Pull request merged**